service/config_provider.go

```diff
@@ -68,50 +68,112 @@ type ConfigProvider interface {
 }
 
 type configProvider struct {
-	locations          []string
-	configMapProviders map[string]config.MapProvider
-	cfgMapConverters   []config.MapConverterFunc
-	configUnmarshaler  configunmarshaler.ConfigUnmarshaler
+	locations           []string
+	configMapProviders  map[string]config.MapProvider
+	configMapConverters []config.MapConverterFunc
+	configUnmarshaler   configunmarshaler.ConfigUnmarshaler
 
 	sync.Mutex
 	closer  config.CloseFunc
 	watcher chan error
 }
 
-// MustNewConfigProvider returns a new ConfigProvider that provides the configuration:
-// * Retrieve the config.Map by merging all retrieved maps from all the config.MapProvider in order.
-// * Then applies all the ConfigMapConverterFunc in the given order.
-// * Then unmarshalls the final config.Config using the given configunmarshaler.ConfigUnmarshaler.
-//
-// The `configMapProviders` is a map of pairs <scheme,Provider>.
+// ConfigProviderOption is an option to change the behavior of ConfigProvider
+// returned by NewConfigProvider()
+type ConfigProviderOption func(opts *configProvider)
+
+// WithConfigMapProvider appends to the default available providers.
+// This provider overwrites any existing provider with the same scheme.
+func WithConfigMapProvider(mp config.MapProvider) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapProviders[mp.Scheme()] = mp
+	}
+}
+
+// WithConfigMapConverters overwrites the default converters.
+func WithConfigMapConverters(c []config.MapConverterFunc) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapConverters = c
```

Comment:
```suggestion
		opts.configMapConverters = append(opts.configMapConverters, c)
```

This should append rather than replace so that multiple `WithConfigMapConverters` options can be composed.

Reply:
The order of the converters matter, so we cannot do append here because will confuse lots of users. The reason why matters is because I may implement a converter that adds a property using an "envvar" then if the env converter is called before or after produces different results.

Reply:
>  we cannot do append here because will confuse lots of users. 

Should `NewConfigProvider` return an error if multiple `WithConfigMapConverters` functions are passed then?

Reply:
Not sure... Personally I don't see a reason to do that.

Reply:
But also this is complicated. We want to have a `default` for this like the expand converter (which probably should not be a converter but that is a separate discussion), if user overwrites that should be accepted correct? Only if it double overwrites is error? Also API accepts slice, so that forces users to think about this...

If you really think that this should be an error let me know, but I don't see this as an error, also the users of this API are very few people :))

Reply:
I think the argument being an array makes it clear enough that folks should set multiple with a single call here.

I can see how it's confusing for `WithConfigMapConverters` to overwrite an option each time it's called while `WithConfigMapProvider` appends to the configuration. I don't have a good idea of how to discern between the two other than with documentation.

Maybe `WithConfigMapProvider` should also overwrite the option to remain consistent?

Reply:
`WithConfigMapProvider` modifies a map, though, where order does not matter and multiple providers may come from multiple sources.  If anything, I'd like to see `WithConfigMapConverters` move in that direction, creating a set of available converters that could then be referenced in a list of conversion steps to apply.  I don't know that there's anything that can be done one way but not the other, perhaps with the exception of having a component with a single method that returns a slice of options intended to register providers and converters.

Reply:
The provider is only one so not sure... We can accept a slice there as well. Also the providers are called in the order of the location argument so their order of registration does not matter.

Another option is to have only one Converter, and add support for MultiConverter as a helper implementation, and then that will overwrite the default converter again

Reply:
@codeboten maybe another good option is o have a "ConfigProviderSettings" instead of using options. And would be clear that we have a map there for providers and a slice for converters. Users can see them easily then?

Reply:
I think that would work. 👍

Reply:
Will give a try in a different PR or commit

Reply:
I think the Settings solution is the best solution to solve the ambiguity that multiple `WithConfigMapConverters` options would bring.

Reply:
Done, settings is available now.